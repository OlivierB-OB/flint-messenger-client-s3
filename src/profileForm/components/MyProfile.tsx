import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import React, { useEffect } from 'react';
import { Action } from 'redux';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { makeResetProfileForm } from '../actions/makeResetProfileForm';
import { makeSaveProfileForm } from '../actions/makeSaveProfileForm';
import { updateProfileForm } from '../actions/updateProfileForm';
import { IProfileFormFields, IProfileFormStatus } from '../types';
import { IdentitySection } from './IdentitySection';
import { CredentialsSection } from './CredentialsSection';
import { IIdentityStatus } from '../../identity/types';
import { Loading } from '../../layout/components/Loading';
import { Alert } from '../../layout/components/Alert';
import { makeDeleteProfile } from '../actions/makeDeleteProfile';

export interface IProfileFormProps {
  identityStatus: IIdentityStatus;
  formStatus: IProfileFormStatus;
  fields: IProfileFormFields;
  update<T extends keyof IProfileFormFields>(field: T, value: IProfileFormFields[T]['value']): void;
  saveProfile(): void;
  resetProfile(): void;
  deleteProfile(): void;
}

export function ProfileForm({
  identityStatus,
  formStatus,
  fields,
  update,
  resetProfile,
  saveProfile,
  deleteProfile,
}: IProfileFormProps) {
  const { email, firstName, lastName, password, confirmation } = fields;
  useEffect(() => void resetProfile(), [identityStatus, resetProfile]);
  if ([identityStatus, formStatus].includes('unavailable')) return <Loading />;
  return (
    <Container>
      <Box style={{ margin: '2rem 0' }}>
        <Alert status={formStatus} success="Profile successfully updated!" />
      </Box>
      <Box style={{ margin: '2rem 0' }}>
        <Grid container justify="flex-end">
          <Grid item xs={2}>
            <Button variant="contained" color="secondary" fullWidth={true} onClick={deleteProfile}>
              Delete account
            </Button>
          </Grid>
        </Grid>
      </Box>
      <form onSubmit={saveProfile}>
        <Box style={{ margin: '2rem 0' }}>
          <Grid container justify="space-evenly" alignItems="flex-start">
            <Grid item xs={4}>
              <IdentitySection
                allowEmailEdition={false}
                email={email}
                firstName={firstName}
                lastName={lastName}
                update={update}
              />
            </Grid>
            <Grid item xs={4}>
              <CredentialsSection password={password} confirmation={confirmation} update={update} />
            </Grid>
          </Grid>
        </Box>
        <Box style={{ margin: '2rem 0' }}>
          <Grid container justify="space-between">
            <Grid item xs={2}>
              <Button variant="contained" fullWidth={true} onClick={resetProfile}>
                Reset
              </Button>
            </Grid>
            <Grid item xs={2}>
              <Button variant="contained" color="primary" fullWidth={true} type="submit">
                Save
              </Button>
            </Grid>
          </Grid>
        </Box>
      </form>
    </Container>
  );
}

const mapStateToProps = ({ identity, profileForm }: IAppState) => ({
  identityStatus: identity.status,
  formStatus: profileForm.status,
  fields: profileForm.fields,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  update: <T extends keyof IProfileFormFields>(field: T, value: IProfileFormFields[T]['value']) =>
    dispatch(updateProfileForm(field, value)),
  resetProfile: () => dispatch(makeResetProfileForm()),
  saveProfile: () => dispatch(makeSaveProfileForm()),
  deleteProfile: () => dispatch(makeDeleteProfile()),
});

export const MyProfile = connect(mapStateToProps, mapDispatchToProps)(ProfileForm);
