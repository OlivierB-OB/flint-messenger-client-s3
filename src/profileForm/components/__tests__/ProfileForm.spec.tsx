import React from 'react';
import { shallow, mount } from 'enzyme';
import { ProfileForm } from '../MyProfile';
import Button from '@material-ui/core/Button';

describe('ProfileForm', () => {
  it('should appear as loading if service is unavailable (1)', async () => {
    const props: any = {
      identityStatus: 'unavailable',
      formStatus: 'ready',
      fields: {},
    };
    const component = shallow(<ProfileForm {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should appear as loading if service is unavailable (2)', async () => {
    const props: any = {
      identityStatus: 'ready',
      formStatus: 'unavailable',
      fields: {},
    };
    const component = shallow(<ProfileForm {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should display the profile form', async () => {
    const props: any = {
      identityStatus: 'ready',
      formStatus: 'the status',
      fields: {
        email: 'email field',
        firstName: 'firstName field',
        lastName: 'lastName field',
        password: 'password field',
        confirmation: 'confirmation field',
      },
      update: 'update',
      resetProfile: 'resetProfile',
      saveProfile: 'saveProfile',
      deleteProfile: 'deleteProfile',
    };
    const component = shallow(<ProfileForm {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should allow deleting', async () => {
    const props: any = {
      identityStatus: 'ready',
      formStatus: 'the status',
      fields: {},
      deleteProfile: jest.fn(),
    };
    const component = shallow(<ProfileForm {...props} />);
    component.find(Button).at(0).simulate('click');
    expect(props.deleteProfile).toHaveBeenCalled();
  });

  it('should allow resetting', async () => {
    const props: any = {
      identityStatus: 'ready',
      formStatus: 'ready',
      fields: {},
      resetProfile: jest.fn(),
    };
    const component = shallow(<ProfileForm {...props} />);
    component.find(Button).at(1).simulate('click');
    expect(props.resetProfile).toHaveBeenCalled();
  });

  it('should allow saving', async () => {
    const props: any = {
      identityStatus: 'ready',
      formStatus: 'ready',
      fields: {},
      saveProfile: jest.fn(),
    };
    const component = shallow(<ProfileForm {...props} />);
    expect(component.find(Button).at(2).prop('type')).toEqual('submit');
    component.find('form').simulate('submit');
    expect(props.saveProfile).toHaveBeenCalled();
  });

  it('should reset the form on mount', async () => {
    const props: any = {
      identityStatus: 'ready',
      formStatus: 'ready',
      fields: {
        email: {
          value: '',
          isValid: true,
        },
        firstName: {
          value: '',
          isValid: true,
        },
        lastName: {
          value: '',
          isValid: true,
        },
        password: {
          value: '',
          isValid: true,
        },
        confirmation: {
          value: '',
          isValid: true,
        },
      },
      update: jest.fn(),
      deleteProfile: jest.fn(),
      resetProfile: jest.fn(),
      saveProfile: jest.fn(),
    };
    mount(<ProfileForm {...props} />);
    expect(props.resetProfile).toHaveBeenCalled();
  });
});
