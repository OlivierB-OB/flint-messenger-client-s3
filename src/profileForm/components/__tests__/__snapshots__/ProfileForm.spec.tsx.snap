// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileForm should appear as loading if service is unavailable (1) 1`] = `<Loading />`;

exports[`ProfileForm should appear as loading if service is unavailable (2) 1`] = `<Loading />`;

exports[`ProfileForm should display the profile form 1`] = `
<WithStyles(ForwardRef(Container))>
  <Styled(MuiBox)
    style={
      Object {
        "margin": "2rem 0",
      }
    }
  >
    <Alert
      status="the status"
      success="Profile successfully updated!"
    />
  </Styled(MuiBox)>
  <Styled(MuiBox)
    style={
      Object {
        "margin": "2rem 0",
      }
    }
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      justify="flex-end"
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={2}
      >
        <WithStyles(ForwardRef(Button))
          color="secondary"
          fullWidth={true}
          onClick="deleteProfile"
          variant="contained"
        >
          Delete account
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </Styled(MuiBox)>
  <form
    onSubmit="saveProfile"
  >
    <Styled(MuiBox)
      style={
        Object {
          "margin": "2rem 0",
        }
      }
    >
      <WithStyles(ForwardRef(Grid))
        alignItems="flex-start"
        container={true}
        justify="space-evenly"
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={4}
        >
          <IdentitySection
            allowEmailEdition={false}
            email="email field"
            firstName="firstName field"
            lastName="lastName field"
            update="update"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={4}
        >
          <CredentialsSection
            confirmation="confirmation field"
            password="password field"
            update="update"
          />
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </Styled(MuiBox)>
    <Styled(MuiBox)
      style={
        Object {
          "margin": "2rem 0",
        }
      }
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        justify="space-between"
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={2}
        >
          <WithStyles(ForwardRef(Button))
            fullWidth={true}
            onClick="resetProfile"
            variant="contained"
          >
            Reset
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={2}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            fullWidth={true}
            type="submit"
            variant="contained"
          >
            Save
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </Styled(MuiBox)>
  </form>
</WithStyles(ForwardRef(Container))>
`;
